react/utils/slots.tsx

```diff
@@ -0,0 +1,179 @@
+import React, { FC, memo, useMemo } from 'react'
+import { useQuery } from 'react-apollo'
+
+import { getChildExtensions } from '../components/ExtensionPoint'
+import ComponentLoader from '../components/ExtensionPoint/ComponentLoader'
+import ListContent from '../queries/ListContent.graphql'
+import { useRuntime } from '../core/main'
+
+interface GenerateSlotArgs {
+  treePath: string
+  slotName: string
+  slotValue: string
+  hydration: Hydration
+}
+
+interface DynamicSlotProps {
+  component: string | null
+  props: Record<string, any>
+  blockId: string | undefined
+  treePath: string
+  runtime: RenderContext
+  hydration: Hydration
+  baseTreePath: string
+}
+
+interface ListContentQuery {
+  listContent: Array<{ contentJSON: string }>
+}
+
+export function generateSlot({
+  treePath,
+  slotName,
+  slotValue,
+  hydration,
+}: GenerateSlotArgs) {
+  const slotMarkerRegex = /(-|#)slot/
+  const slotMarkerInTreePath = treePath.match(slotMarkerRegex)
+  const treePathWithoutSlotMarker = slotMarkerInTreePath
+    ? treePath.substring(0, treePath.indexOf(slotMarkerInTreePath[0]))
+    : treePath
+
+  const newTreePath = `${treePathWithoutSlotMarker}/${slotValue}`
+
+  const SlotComponent: FC<any> = memo(props => {
+    const runtime = useRuntime()
+    let extension = runtime.extensions[newTreePath]
+
+    const slotChildren = getChildExtensions(runtime, newTreePath)
+    let componentProps = extension?.props ?? {}
+
+    if (props.id) {
+      const hasLabel = slotValue.includes('#')
+      const dynamicTreePath = `${newTreePath}${hasLabel ? '-' : '#'}slot${
+        props.id
+      }`
+
+      if (!(dynamicTreePath in runtime.extensions)) {
```

Comment:
What does that mean? Mind adding comments too?

Another way to comment the code is to assign this condition to a variable with a good name:

```ts
const slotHasNoSettings = !(dynamicTreePath in runtime.extensions)
if (slotHasNoSettings) {
```